File: rust_mlp/src/mlp.rs
// abstract implementation of mlp, with layers and activation functions
// concretization is meant to be done in main

use crate::loss::LossFunction;
use crate::layer::{DenseLayer, Layer};
use num_traits::real::Real;

pub trait MLP<T: Real> {
    fn forward(&self, input: &Vec<T>) -> Vec<Vec<T>>;
    fn backward(&mut self, input: &Vec<Vec<T>>, input: &Vec<T>, target: &Vec<T>) -> (Vec<Vec<T>>, Vec<Vec<Vec<T>>>);
}

pub struct MultiLayerPerceptron<T: Real> {
    pub layers: Vec<DenseLayer<T>>,
    pub loss: Box<dyn LossFunction<T>>,
}

impl<T: Real> MultiLayerPerceptron<T> {
    pub fn new(loss: impl LossFunction<T> + 'static) -> Self {
        MultiLayerPerceptron {
            layers: Vec::new(),
            loss: Box::new(loss),
        }
    }

    pub fn add_layer(&mut self, layer: DenseLayer<T>) {
        self.layers.push(layer);
    }
}

impl<T: Real> MLP<T> for MultiLayerPerceptron<T> {
    fn forward(&self, input: &Vec<T>) -> Vec<Vec<T>> {
        let mut layer_outputs: Vec<Vec<T>> = Vec::new();
        let mut current_input = input.clone();
        for layer in &self.layers {
            let output = layer.forward(&current_input);
            layer_outputs.push(output.clone());
            current_input = output;
        }
        layer_outputs
    }

    fn backward(&mut self, layer_outputs: &Vec<Vec<T>>, input: &Vec<T>, target: &Vec<T>) -> (Vec<Vec<T>>, Vec<Vec<Vec<T>>>) {
        // first, calculate deltas for the last layer
        let mut deltas_matrix: Vec<Vec<T>> = Vec::new();
        let mut gradients_matrix: Vec<Vec<Vec<T>>> = Vec::new();
        let mut deltas: Vec<T> = self.loss.derivative(&layer_outputs[layer_outputs.len() - 1], &target);
        let mut gradients: Vec<Vec<T>>;
        // then, calculate deltas for the other layers
        for i in (0..self.layers.len()).rev() {
            let layer_input = if i == 0 {
                input.clone()
            } else {
                layer_outputs[i - 1].clone()
            };
            let next_weights = if i == self.layers.len() - 1 {
                Vec::new()
            } else {
                self.layers[i + 1].weights.clone()
            };
            (deltas, gradients) = self.layers[i].backward(&layer_input, &layer_outputs[i], &next_weights, &deltas);
            deltas_matrix.push(deltas.clone());
            gradients_matrix.push(gradients);

        }
        deltas_matrix.reverse();
        gradients_matrix.reverse();
        (deltas_matrix, gradients_matrix) // return the deltas and gradients for each layer to the optimizer
    }
}

File: py_mlp/mlp.py
import numpy as np
from loss import LossFunction
from layer import Layer
from typing import List, Tuple

class MultiLayerPerceptron:
    layers: List[Layer]
    loss: LossFunction

    def __init__(self, loss: LossFunction):
        self.layers = []
        self.loss = loss

    def __call__(self, x: np.ndarray) -> np.ndarray:
        return self.predict(x)

    def add_layer(self, layer: Layer):
        self.layers.append(layer)

    def predict(self, x: np.ndarray) -> np.ndarray:
        for layer in self.layers:
            x = layer(x)
        return x

    def forward(self, x: np.ndarray) -> np.ndarray:
        return [x := layer(x) for layer in self.layers]

    def backward(self, layer_outputs: List[np.ndarray], input: np.ndarray, target: np.ndarray) -> List[np.ndarray]:
        """
        Returns the gradients back to the optimizer
        """
        deltas = [self.loss.gradient(target, layer_outputs[-1])]
        gradients = []
        for i in range(len(self.layers) - 1, -1, -1):
            layer_input = layer_outputs[i - 1] if i > 0 else input
            next_weights = self.layers[i + 1].weights if i < len(self.layers) - 1 else None
            new_deltas, new_gradients = self.layers[i].backward(layer_input, layer_outputs[i], deltas[-1], next_weights)
            deltas.append(new_deltas)
            gradients.append(new_gradients)
        return deltas[::-1], gradients[::-1]

    def detect_nan(self):
        for layer in self.layers:
            if np.isnan(layer.weights).any() or np.isnan(layer.biases).any():
                raise ValueError("NaN detected in weights or biases")